freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java

```diff
@@ -42,11 +44,34 @@ public void validateStackForVerticalUpscale(Stack stack, VerticalScaleRequest re
         verticalScalingValidatorService.validateRequest(stack, request);
     }
 
-    public void validateStackForDownscale(Set<InstanceMetaData> allInstances, Stack stack, ScalingPath scalingPath) {
+    public void validateStackForDownscale(Set<InstanceMetaData> allInstances, Stack stack, ScalingPath scalingPath, Set<String> instanceIdsToDelete) {
         validateScalingIsDownscale(scalingPath);
+        validateInstanceIdsToDelete(allInstances, instanceIdsToDelete);
         executeCommonValidations(allInstances, stack, scalingPath, OperationType.DOWNSCALE);
     }
 
+    private void validateInstanceIdsToDelete(Set<InstanceMetaData> allInstances, Set<String> instanceIdsToDelete) {
+        if (Objects.nonNull(instanceIdsToDelete)) {
+            validateInstanceIdsAreNotEmpty(instanceIdsToDelete);
+            validateInstanceIdsArePartOfAllInstances(allInstances, instanceIdsToDelete);
+            validateInstanceIdToDeleteAreNotPrimaryGateways(allInstances, instanceIdsToDelete);
+        }
+    }
+
+    private void validateInstanceIdToDeleteAreNotPrimaryGateways(Set<InstanceMetaData> allInstances, Set<String> instanceIds) {
```

Comment:
why is this a requirement?

Reply:
Previously, when [ freeipa calculated which nodes can be deleted](https://github.com/hortonworks/cloudbreak/blob/354d0c54a3dba1a2ec6259af61d13531c8f6d958/freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingService.java#L161), it only took non-primary gateway nodes. I thus converted it into a validation. 
Do we want to allow primary gateways to be downscaled? We would need to convert the downscale to a flow chain and add change primary gateway flow in that case.

Reply:
ok, I think we can go with this version for now and we can have a followup ticket/pr if this causes any issues